config/jobs/kubernetes/sig-node/dynamic-resource-allocation-canary.yaml

```diff
@@ -0,0 +1,320 @@
+# This file contains canary periodic and presubmit jobs which run tests covering
+# Dynamic Resource Allocation (DRA).
+#
+# The intent is to make all changes to DRA jobs first for these canary jobs,
+# then copy the changes into dynamic-resource-allocation.yaml. Unless some
+# experimental changes are being tested, a diff between the two files should
+# be limited to:
+# - this comment
+# - job names
+# - interval for the periodic jobs
+#
+# This command can be used to check this:
+#    diff dynamic-resource-allocation.yaml <(sed -e 's/-canary//' dynamic-resource-allocation-canary.yaml)
+
+
+# `templates` has no special meaning. It just holds YAML anchors (= re-usable content)
+# that get referenced below via YAML aliases: `<<: *job` includes the content
+# and then allows adding or overwriting fields. This is done at the root. If a field
+# contains lists or objects, that content gets replaced instead of merged.
+#
+# Lists cannot be extended the same way (https://github.com/yaml/yaml/issues/35).
+#
+# If unsure what the expanded jobs look like or to test parsing, run `go test -v .`
```

Comment:
@kannon92 @bart0sh: I went ahead and rewrote this file so that it avoids duplication between the different jobs. From 3942639dbf072b7c272b21039dae59dc9bf10988:

    dra: use YAML anchors and aliases to avoid duplication
    
    These different jobs were all similar, with small variations.
    Keeping them in sync was an on-going challenge that we lost:
    - at some point, features were enabled differently in E2E node
      presubmits and periodics (didn't make a difference, but the
      they weren't the same anymore)
    - resource settings for containerd vs CRI-O were different
      and it is unclear whether that was intentional (no comment
      about it)
    - clusters where different
    
    Defining common elements via YAML anchors once and reusing them via YAML
    aliases avoids this.
    
    `go test -v -run=TestYaml/dynamic-resource-allocation-canary.yaml .`
    can be used to see the full job definitions.

What do you think?

I know @bart0sh pointed out the difficulty of keeping periodics and presubmits in sync. This approach solves that, but is it readable and usable enough?

Reply:
There are some test failures that would need further work. But let's first hear whether this is worth it at all.

Perhaps there is some other templating mechanism that can be used instead to generate the YAML files?

Reply:
This looks better than what we currently have.
However, I'd prefer to use [existing solution](https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes/kops/templates) suggested by @dims [on the slack](https://app.slack.com/client/T09NY5SBT/C09QZ4DQB).

Reply:
@pohly It would be even better to avoid having a lot of configuration details in the generation script (`https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/kops/build_jobs.py`) and have a yaml file with a set of values, templates for various job types and a generic build_jobs script that generates result by rendering templates based on values.yaml. I can play with this if you don't have time.

Reply:
I agree, this PR is going down the wrong track. Instead of two files which we need to keep in sync, we should have a single `dynamic-resource-allocation.yaml` which is fully generated. That file should also include canary presubmits where we can try out experimental changes before changing the generator to apply them also to the normal jobs.

We don't need canary periodic jobs, for two reasons:
- If the difference between presubmit and periodic is guaranteed to be limited to the `interval` vs. `branch` settings, then testing the job spec with a canary presubmit is sufficient.
- It is less critical if a periodic job breaks because it doesn't affect other developers in the PRs.

> I can play with this if you don't have time.

Thanks, please do.

Reply:
@pohly this is what I came up with so far, PTAL: https://github.com/kubernetes/test-infra/pull/34010